[Bug]: Environment.report(...) throws NullPointerException

### Describe the bug

using getEnvironment.report(processor, Level.INFO, element, "Message...") in my processor i get a NullPointerException if the element is implicit.


### Source code you are trying to analyze/transform

```Java
package fr.inria.gforge.spoon.analysis;

/**
 * The Class Testcase2.
 *
 * testcase to show NullPointerException
 */
public class Testcase2
{

    private class A
    {

    }

    private class B extends A
    {

        /**
         * Instantiates a new b.
         */
        public B()
        {
            super();
        }
    }

    public static void main(String[] args)
    {
        Testcase2 t = new Testcase2();
        t.test();
    }

    public String test()
    {
        var b = new B();
        return b.toString();
    }
}
```


### Source code for your Spoon processing

```Java
package fr.inria.gforge.spoon.analysis;

import java.util.List;

import spoon.processing.AbstractProcessor;
import spoon.reflect.code.CtInvocation;
import spoon.reflect.declaration.CtConstructor;
import spoon.reflect.visitor.Query;
import spoon.reflect.visitor.filter.TypeFilter;
import spoon.support.Level;

/**
 * Processor to show {@link NullPointerException} thrown if element is implicit.
 *
 * @author r.b.
 *
 */
public class BugProcessor extends AbstractProcessor<CtConstructor<?>>
{

    /**
     * implementation walks through all invocations and just prints them
     *
     * @see spoon.processing.AbstractProcessor#isToBeProcessed(spoon.reflect.declaration.CtElement)
     *
     * @param candidate element to check
     * @return true if the element should be processed
     */
    public boolean isToBeProcessed(CtConstructor<?> candidate)
    {

        boolean process = false;
        List<CtInvocation<?>> invocations = Query.getElements(candidate, new TypeFilter<>(CtInvocation.class));

        invocations.forEach(i -> getEnvironment().report(this, Level.INFO, i, "Message"));

        return process;
    }

    /**
     * @see spoon.processing.Processor#process(spoon.reflect.declaration.CtElement)
     *
     * @param element
     */
    @Override
    public void process(CtConstructor<?> element)
    {
        // no functionality here ...
    }
}
```


### Actual output

```Java
Caused by: java.lang.NullPointerException
    at spoon.support.StandardEnvironment.report (StandardEnvironment.java:265)
    at fr.inria.gforge.spoon.analysis.BugProcessor.lambda$isToBeProcessed$0 (BugProcessor.java:57)
    at java.util.ArrayList.forEach (ArrayList.java:1541)
    at fr.inria.gforge.spoon.analysis.BugProcessor.isToBeProcessed (BugProcessor.java:57)
    at fr.inria.gforge.spoon.analysis.BugProcessor.isToBeProcessed (BugProcessor.java:40)
    at spoon.support.visitor.ProcessingVisitor.scan (ProcessingVisitor.java:71)
```


### Expected output

```Java
No Exception
```


### Spoon Version

10.2.0

### JVM Version

11

### What operating system are you using?

win10

--- a/src/main/java/spoon/support/StandardEnvironment.java
+++ b/src/main/java/spoon/support/StandardEnvironment.java
@@ -256,13 +256,20 @@ public class StandardEnvironment implements Serializable, Environment {
 		if (sp == null) {
 			buffer.append(" (Unknown Source)");
 		} else {
-			// TODO: will explode if type == null
-			buffer.append(" at " + type.getQualifiedName() + ".");
+			if (type != null) {
+				buffer.append(" at " + type.getQualifiedName() + ".");
+			} else {
+				buffer.append("at  (??).");
+			}
 			CtExecutable<?> exe = (element instanceof CtExecutable) ? (CtExecutable<?>) element : element.getParent(CtExecutable.class);
 			if (exe != null) {
 				buffer.append(exe.getSimpleName());
 			}
-			buffer.append("(" + sp.getFile().getName() + ":" + sp.getLine() + ")");
+			if (sp.getFile() != null) {
+				buffer.append("(" + sp.getFile().getName() + ":" + sp.getLine() + ")");
+			} else {
+				buffer.append("(?:?)");
+			}
 		}
 
 		print(buffer.toString(), level);

--- a/src/main/java/spoon/support/StandardEnvironment.java
+++ b/src/main/java/spoon/support/StandardEnvironment.java
@@ -257,16 +257,22 @@ public class StandardEnvironment implements Serializable, Environment {
 			buffer.append(" (Unknown Source)");
 		} else {
 			if (type != null) {
-				buffer.append(" at " + type.getQualifiedName() + ".");
+				buffer.append(" at ")
+							.append(type.getQualifiedName())
+							.append(".");
 			} else {
-				buffer.append("at  (??).");
+				buffer.append("at (?).");
 			}
 			CtExecutable<?> exe = (element instanceof CtExecutable) ? (CtExecutable<?>) element : element.getParent(CtExecutable.class);
 			if (exe != null) {
 				buffer.append(exe.getSimpleName());
 			}
 			if (sp.getFile() != null) {
-				buffer.append("(" + sp.getFile().getName() + ":" + sp.getLine() + ")");
+				buffer.append("(")
+							.append(sp.getFile().getName())
+							.append(":")
+							.append(sp.getLine())
+							.append(")");
 			} else {
 				buffer.append("(?:?)");
 			}
@@ -274,6 +280,7 @@ public class StandardEnvironment implements Serializable, Environment {
 
 		print(buffer.toString(), level);
 	}
+
 
 	@Override
 	public void report(Processor<?> processor, Level level, String message) {

--- a/src/test/java/spoon/test/logging/LogTest.java
+++ b/src/test/java/spoon/test/logging/LogTest.java
@@ -20,15 +20,23 @@ import org.apache.commons.lang3.tuple.Pair;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.provider.MethodSource;
 import org.junit.jupiter.params.ParameterizedTest;
+import spoon.FluentLauncher;
 import spoon.Launcher;
 import spoon.MavenLauncher;
+import spoon.processing.AbstractProcessor;
+import spoon.reflect.CtModel;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.declaration.CtConstructor;
+import spoon.reflect.visitor.Query;
+import spoon.reflect.visitor.filter.TypeFilter;
 import spoon.support.JavaOutputProcessor;
 import spoon.support.Level;
+import spoon.test.GitHubIssue;
 import uk.org.lidalia.slf4jtest.TestLogger;
 import uk.org.lidalia.slf4jtest.TestLoggerFactory;
-
+import java.util.List;
 import java.util.stream.Stream;
-
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
@@ -102,5 +110,23 @@ public class LogTest {
 					"This is a message with level " + level.toString());
 		}
 		assertEquals(0, logger.getLoggingEvents().size());
+	}
+
+	@GitHubIssue(issueNumber = 4997, fixed = true)
+	void innerTypesCrashesLogging() {
+		// contract: when a class has inner types, the logging should not crash with a NPE
+		String codePath = "src/test/resources/logging/TestCase2.java";
+		var processor = new AbstractProcessor<CtConstructor<?>>() {
+			@Override
+			public void process(CtConstructor<?> element) {
+				// do nothing
+			}
+			public boolean isToBeProcessed(CtConstructor<?> candidate) {
+        List<CtInvocation<?>> invocations = Query.getElements(candidate, new TypeFilter<>(CtInvocation.class));
+        invocations.forEach(i -> getEnvironment().report(this, Level.INFO, i, "Message"));
+        return false;
+			}
+		};
+		assertDoesNotThrow(() ->new FluentLauncher().inputResource(codePath).processor(processor).buildModel());
 	}
 }

--- a/src/test/java/spoon/test/logging/LogTest.java
+++ b/src/test/java/spoon/test/logging/LogTest.java
@@ -121,12 +121,11 @@ public class LogTest {
 			public void process(CtConstructor<?> element) {
 				// do nothing
 			}
-			public boolean isToBeProcessed(CtConstructor<?> candidate) {
-        List<CtInvocation<?>> invocations = Query.getElements(candidate, new TypeFilter<>(CtInvocation.class));
-        invocations.forEach(i -> getEnvironment().report(this, Level.INFO, i, "Message"));
-        return false;
-			}
-		};
-		assertDoesNotThrow(() ->new FluentLauncher().inputResource(codePath).processor(processor).buildModel());
+		public boolean isToBeProcessed(CtConstructor<?> candidate) {
+        		List<CtInvocation<?>> invocations = Query.getElements(candidate, new TypeFilter<>(CtInvocation.class));
+        		invocations.forEach(i -> getEnvironment().report(this, Level.INFO, i, "Message"));
+       				return false;
+	  	}};
+		assertDoesNotThrow(() -> new FluentLauncher().inputResource(codePath).processor(processor).buildModel());
 	}
 }

--- a/src/test/java/spoon/test/logging/LogTest.java
+++ b/src/test/java/spoon/test/logging/LogTest.java
@@ -121,11 +121,11 @@ public class LogTest {
 			public void process(CtConstructor<?> element) {
 				// do nothing
 			}
-		public boolean isToBeProcessed(CtConstructor<?> candidate) {
-        		List<CtInvocation<?>> invocations = Query.getElements(candidate, new TypeFilter<>(CtInvocation.class));
-        		invocations.forEach(i -> getEnvironment().report(this, Level.INFO, i, "Message"));
+			public boolean isToBeProcessed(CtConstructor<?> candidate) {
+        			List<CtInvocation<?>> invocations = Query.getElements(candidate, new TypeFilter<>(CtInvocation.class));
+        			invocations.forEach(i -> getEnvironment().report(this, Level.INFO, i, "Message"));
        				return false;
-	  	}};
+	  		}};
 		assertDoesNotThrow(() -> new FluentLauncher().inputResource(codePath).processor(processor).buildModel());
 	}
 }

--- a/src/test/resources/logging/TestCase2.java
+++ b/src/test/resources/logging/TestCase2.java
@@ -1,0 +1,33 @@
+package fr.inria.gforge.spoon.analysis;
+
+/**
+ * The Class Testcase2.
+ *
+ * testcase to show NullPointerException
+ */
+public class Testcase2 {
+
+  private class A {
+
+  }
+
+  private class B extends A {
+
+    /**
+     * Instantiates a new b.
+     */
+    public B() {
+      super();
+    }
+  }
+
+  public static void main(String[] args) {
+    Testcase2 t = new Testcase2();
+    t.test();
+  }
+
+  public String test() {
+    var b = new B();
+    return b.toString();
+  }
+}

